File: Data_Structure/Linear_List/C++_Version/code.cpp
#include <iostream>
#include <exception>
using namespace std;

// 异常类定义
class outOfRange : public exception {
public:
    const char* what() const throw() { return "ERROR! OUT OF RANGE.\n"; }
};

class badSize : public exception {
public:
    const char* what() const throw() { return "ERROR! BAD SIZE.\n"; }
};

// 抽象基类
template<class T>
class List {
public:
    virtual void clear() = 0;
    virtual bool empty() const = 0;
    virtual int size() const = 0;
    virtual void insert(int i, const T &value) = 0;
    virtual void remove(int i) = 0;
    virtual int search(const T &value) const = 0;
    virtual T visit(int i) const = 0;
    virtual void traverse() const = 0;
    virtual void inverse() = 0;
    virtual ~List() {}
};

// 顺序表实现
template <class elemType>
class seqList : public List<elemType> {
private:
    elemType *data;
    int curLength;
    int maxSize;
    void resize() {
        elemType* p = data;
        maxSize *= 2;
        data = new elemType[maxSize];
        for (int i = 0; i < curLength; i++)
            data[i] = p[i];
        delete[] p;
    }
public:
    seqList(int initSize = 100) {
        if (initSize <= 0) throw badSize();
        maxSize = initSize;
        data = new elemType[maxSize];
        curLength = 0;
    }

    seqList(const seqList& sl) {
        maxSize = sl.maxSize;
        curLength = sl.curLength;
        data = new elemType[maxSize];
        for (int i = 0; i < curLength; ++i)
            data[i] = sl.data[i];
    }

    ~seqList() { delete[] data; }

    void clear() { curLength = 0; }
    bool empty() const { return curLength == 0; }
    int size() const { return curLength; }

    void traverse() const {
        if (empty()) cout << "is empty" << endl;
        else {
            cout << "output element: " << endl;
            for (int i = 0; i < curLength; i++)
                cout << data[i] << " ";
            cout << endl;
        }
    }

    int search(const elemType& value) const {
        for (int i = 0; i < curLength; i++)
            if (value == data[i]) return i;
        return -1;
    }

    void insert(int i, const elemType& value) {
        if (i < 0 || i > curLength) throw outOfRange();
        if (curLength == maxSize) resize();
        for (int j = curLength; j > i; j--)
            data[j] = data[j-1];
        data[i] = value;
        ++curLength;
    }

    void remove(int i) {
        if (i < 0 || i > curLength - 1) throw outOfRange();
        for (int j = i; j < curLength - 1; j++)
            data[j] = data[j+1];
        --curLength;
    }

    elemType visit(int i) const {
        if (i < 0 || i >= curLength) throw outOfRange();
        return data[i];
    }

    void inverse() {
        elemType tmp;
        for (int i = 0; i < curLength/2; i++) {
            tmp = data[i];
            data[i] = data[curLength-1-i];
            data[curLength-1-i] = tmp;
        }
    }
};

// 单链表节点结构
template <class elemType>
struct Node {
    elemType data;
    Node* next;
    Node(const elemType value, Node* p = nullptr) : data(value), next(p) {}
    Node(Node* p = nullptr) : next(p) {}
};

// 单链表实现
template <class elemType>
class linkList : public List<elemType> {
private:
    struct Node {
        elemType data;
        Node* next;
        Node(const elemType value, Node* p = nullptr) : data(value), next(p) {}
        Node(Node* p = nullptr) : next(p) {}
    };
    Node* head;
    Node* tail;
    int curLength;

    Node* getPosition(int i) const {
        if (i < -1 || i > curLength-1) return nullptr;
        Node* p = head;
        int count = 0;
        while (count <= i) {
            p = p->next;
            count++;
        }
        return p;
    }

public:
    linkList() {
        head = tail = new Node();
        curLength = 0;
    }

    ~linkList() { clear(); delete head; }

    void clear() {
        Node *p = head->next, *tmp;
        while (p) {
            tmp = p;
            p = p->next;
            delete tmp;
        }
        head->next = nullptr;
        tail = head;
        curLength = 0;
    }

    bool empty() const { return head->next == nullptr; }
    int size() const { return curLength; }

    void traverse() const {
        Node* p = head->next;
        while (p) {
            cout << p->data << " ";
            p = p->next;
        }
        cout << endl;
    }

    int search(const elemType& value) const {
        Node* p = head->next;
        int count = 0;
        while (p && p->data != value) {
            p = p->next;
            count++;
        }
        return p ? count : -1;
    }

    void insert(int i, const elemType& value) {
        if (i < 0 || i > curLength) throw outOfRange();
        Node* p = getPosition(i-1);
        Node* q = new Node(value, p->next);
        p->next = q;
        if (p == tail) tail = q;
        curLength++;
    }

    void remove(int i) {
        if (i < 0 || i > curLength-1) throw outOfRange();
        Node* pre = getPosition(i-1);
        Node* p = pre->next;
        if (p == tail) {
            tail = pre;
            pre->next = nullptr;
        } else {
            pre->next = p->next;
        }
        delete p;
        curLength--;
    }

    elemType visit(int i) const {
        if (i < 0 || i >= curLength) throw outOfRange();
        Node* p = head->next;
        for (int count = 0; count < i; count++)
            p = p->next;
        return p->data;
    }

    void inverse() {
        Node* p = head->next;
        head->next = nullptr;
        if (p) tail = p;
        while (p) {
            Node* tmp = p->next;
            p->next = head->next;
            head->next = p;
            p = tmp;
        }
    }
};

// 双链表节点结构
template <class elemType>
struct DoubleNode {
    elemType data;
    DoubleNode* prior;
    DoubleNode* next;
    DoubleNode(const elemType value, DoubleNode* p = nullptr, DoubleNode* n = nullptr)
        : data(value), prior(p), next(n) {}
    DoubleNode() : next(nullptr), prior(nullptr) {}
};

// 双链表实现
template <class elemType>
class doubleLinkList : public List<elemType> {
private:
    struct DoubleNode {
        elemType data;
        DoubleNode* prior;
        DoubleNode* next;
        DoubleNode(const elemType value, DoubleNode* p = nullptr, DoubleNode* n = nullptr)
            : data(value), prior(p), next(n) {}
        DoubleNode() : next(nullptr), prior(nullptr) {}
    };
    DoubleNode* head;
    DoubleNode* tail;
    int curLength;

    DoubleNode* getPosition(int i) const {
        DoubleNode* p = head;
        int count = 0;
        if (i < -1 || i > curLength) return nullptr;
        while (count <= i) {
            p = p->next;
            count++;
        }
        return p;
    }

public:
    doubleLinkList() {
        head = new DoubleNode;
        tail = new DoubleNode;
        head->next = tail;
        tail->prior = head;
        curLength = 0;
    }

    ~doubleLinkList() {
        clear();
        delete head;
        delete tail;
    }

    void clear() {
        DoubleNode* p = head->next;
        DoubleNode* tmp;
        head->next = tail;
        tail->prior = head;
        while (p != tail) {
            tmp = p->next;
            delete p;
            p = tmp;
        }
        curLength = 0;
    }

    bool empty() const { return head->next == tail; }
    int size() const { return curLength; }

    void traverse() const {
        DoubleNode* p = head->next;
        while (p != tail) {
            cout << p->data << " ";
            p = p->next;
        }
        cout << endl;
    }

    int search(const elemType& value) const {
        DoubleNode* p = head->next;
        int count = 0;
        while (p != tail && p->data != value) {
            p = p->next;
            count++;
        }
        return (p != tail) ? count : -1;
    }

    void insert(int i, const elemType& value) {
        if (i < 0 || i > curLength) throw outOfRange();
        DoubleNode* p = getPosition(i);
        DoubleNode* tmp = new DoubleNode(value, p->prior, p);
        p->prior->next = tmp;
        p->prior = tmp;
        ++curLength;
    }

    void remove(int i) {
        if (i < 0 || i > curLength-1) throw outOfRange();
        DoubleNode* p = getPosition(i);
        p->prior->next = p->next;
        p->next->prior = p->prior;
        delete p;
        --curLength;
    }

    elemType visit(int i) const {
        if (i < 0 || i >= curLength) throw outOfRange();
        DoubleNode* p = head->next;
        for (int count = 0; count < i; count++)
            p = p->next;
        return p->data;
    }

    void inverse() {
        DoubleNode* tmp, *p = head->next;
        head->next = tail;
        tail->prior = head;
        while (p != tail) {
            tmp = p->next;
            p->next = head->next;
            p->prior = head;
            head->next->prior = p;
            head->next = p;
            p = tmp;
        }
    }
};

// 测试示例
int main() {
    // 顺序表测试
    seqList<int> sl(5);
    for(int i=0; i<5; i++) sl.insert(i, i*2);
    cout << "顺序表初始: "; sl.traverse();
    sl.inverse();
    cout << "顺序表逆置: "; sl.traverse();

    // 单链表测试
    linkList<int> ll;
    for(int i=0; i<5; i++) ll.insert(i, i+1);
    cout << "单链表初始: "; ll.traverse();
    ll.remove(2);
    cout << "删除第三个元素: "; ll.traverse();

    // 双链表测试
    doubleLinkList<int> dl;
    for(int i=0; i<3; i++) dl.insert(i, i*3);
    cout << "双链表初始: "; dl.traverse();
    dl.inverse();
    cout << "双链表逆置: "; dl.traverse();

    return 0;
}

File: Data_Structure/Linear_List/Python_Version/code.py
from abc import ABC, abstractmethod
from typing import TypeVar, Generic, Optional, Any

T = TypeVar('T')

# 异常类定义 - 对应C++的outOfRange和badSize
class OutOfRange(Exception):
    """对应C++的outOfRange异常"""
    def __str__(self):
        return "ERROR! OUT OF RANGE."

class BadSize(Exception):
    """对应C++的badSize异常"""
    def __str__(self):
        return "ERROR! BAD SIZE."

# 抽象基类 - 对应C++的List模板类
class List(ABC, Generic[T]):
    """线性表抽象基类，定义所有线性表必须实现的接口"""
    
    @abstractmethod
    def clear(self) -> None:
        """清空线性表"""
        pass
    
    @abstractmethod
    def empty(self) -> bool:
        """判断线性表是否为空"""
        pass
    
    @abstractmethod
    def size(self) -> int:
        """返回线性表长度"""
        pass
    
    @abstractmethod
    def insert(self, i: int, value: T) -> None:
        """在位置i插入元素value"""
        pass
    
    @abstractmethod
    def remove(self, i: int) -> None:
        """删除位置i的元素"""
        pass
    
    @abstractmethod
    def search(self, value: T) -> int:
        """查找元素value，返回位置索引，未找到返回-1"""
        pass
    
    @abstractmethod
    def visit(self, i: int) -> T:
        """访问位置i的元素"""
        pass
    
    @abstractmethod
    def traverse(self) -> None:
        """遍历输出所有元素"""
        pass
    
    @abstractmethod
    def inverse(self) -> None:
        """逆置线性表"""
        pass

# 顺序表实现 - 对应C++的seqList
class SeqList(List[T]):
    """顺序表实现，使用动态数组存储"""
    
    def __init__(self, init_size: int = 100):
        """构造函数，对应C++构造函数"""
        if init_size <= 0:
            raise BadSize()
        self._max_size = init_size
        self._data: list[Optional[T]] = [None] * self._max_size  # 对应C++的elemType* data
        self._cur_length = 0  # 对应C++的curLength
    
    def _resize(self) -> None:
        """动态扩容，对应C++的resize()私有方法"""
        old_data = self._data
        self._max_size *= 2
        self._data: list[Optional[T]] = [None] * self._max_size
        for i in range(self._cur_length):
            self._data[i] = old_data[i]
    
    def clear(self) -> None:
        """清空顺序表"""
        self._cur_length = 0
    
    def empty(self) -> bool:
        """判断是否为空"""
        return self._cur_length == 0
    
    def size(self) -> int:
        """返回当前长度"""
        return self._cur_length
    
    def traverse(self) -> None:
        """遍历输出，格式与C++完全一致"""
        if self.empty():
            print("is empty")
        else:
            for i in range(self._cur_length):
                print(self._data[i], end=" ")
            print()
    
    def search(self, value: T) -> int:
        """查找元素，返回第一个匹配位置"""
        for i in range(self._cur_length):
            if value == self._data[i]:
                return i
        return -1
    
    def insert(self, i: int, value: T) -> None:
        """在位置i插入元素value"""
        if i < 0 or i > self._cur_length:
            raise OutOfRange()
        if self._cur_length == self._max_size:
            self._resize()
        
        # 元素后移，对应C++的循环
        for j in range(self._cur_length, i, -1):
            self._data[j] = self._data[j-1]
        
        self._data[i] = value
        self._cur_length += 1
    
    def remove(self, i: int) -> None:
        """删除位置i的元素"""
        if i < 0 or i > self._cur_length - 1:
            raise OutOfRange()
        
        # 元素前移，对应C++的循环
        for j in range(i, self._cur_length - 1):
            self._data[j] = self._data[j+1]
        
        self._cur_length -= 1
    
    def visit(self, i: int) -> T:
        """访问位置i的元素"""
        if i < 0 or i >= self._cur_length:
            raise OutOfRange()
        result = self._data[i]
        if result is None:
            raise RuntimeError("Unexpected None value in valid position")
        return result
    
    def inverse(self) -> None:
        """逆置顺序表，使用交换方式"""
        for i in range(self._cur_length // 2):
            # 对应C++中的tmp交换
            self._data[i], self._data[self._cur_length-1-i] = \
                self._data[self._cur_length-1-i], self._data[i]

# 单链表节点 - 对应C++的Node结构
class _LinkNode(Generic[T]):
    """单链表节点，对应C++的Node结构"""
    def __init__(self, data: Optional[T] = None, next_node: Optional['_LinkNode[T]'] = None):
        self.data = data  # 对应C++的data成员
        self.next = next_node  # 对应C++的next指针

# 单链表实现 - 对应C++的linkList
class LinkList(List[T]):
    """单链表实现，带头节点和尾指针"""
    
    def __init__(self):
        """构造函数，创建头节点和尾指针"""
        self._head = _LinkNode[T]()  # 对应C++的head，头节点
        self._tail = self._head  # 对应C++的tail，尾指针
        self._cur_length = 0  # 对应C++的curLength
    
    def _get_position(self, i: int) -> Optional[_LinkNode[T]]:
        """获取位置i的前一个节点，对应C++的getPosition()"""
        if i < -1 or i > self._cur_length - 1:
            return None
        
        p = self._head
        count = 0
        while count <= i and p is not None:
            p = p.next
            count += 1
        return p
    
    def clear(self) -> None:
        """清空链表，对应C++的clear()"""
        p = self._head.next
        while p is not None:
            tmp = p
            p = p.next
            del tmp  # 显式删除节点
        
        self._head.next = None
        self._tail = self._head
        self._cur_length = 0
    
    def empty(self) -> bool:
        """判断是否为空"""
        return self._head.next is None
    
    def size(self) -> int:
        """返回链表长度"""
        return self._cur_length
    
    def traverse(self) -> None:
        """遍历输出链表"""
        p = self._head.next
        while p is not None:
            print(p.data, end=" ")
            p = p.next
        print()
    
    def search(self, value: T) -> int:
        """查找元素value"""
        p = self._head.next
        count = 0
        while p is not None and p.data != value:
            p = p.next
            count += 1
        return count if p is not None else -1
    
    def insert(self, i: int, value: T) -> None:
        """在位置i插入元素value"""
        if i < 0 or i > self._cur_length:
            raise OutOfRange()
        
        # 找到插入位置的前一个节点
        if i == 0:
            p = self._head
        else:
            p = self._get_position(i-1)
            if p is None:
                raise OutOfRange()
        
        # 创建新节点并插入
        q = _LinkNode(value, p.next)
        p.next = q
        
        # 更新尾指针
        if p == self._tail:
            self._tail = q
        
        self._cur_length += 1
    
    def remove(self, i: int) -> None:
        """删除位置i的元素"""
        if i < 0 or i > self._cur_length - 1:
            raise OutOfRange()
        
        # 找到删除位置的前一个节点
        if i == 0:
            pre = self._head
        else:
            pre = self._get_position(i-1)
            if pre is None or pre.next is None:
                raise OutOfRange()
        
        p = pre.next
        if p is None:
            raise OutOfRange()
        
        # 更新尾指针
        if p == self._tail:
            self._tail = pre
            pre.next = None
        else:
            pre.next = p.next
        
        del p
        self._cur_length -= 1
    
    def visit(self, i: int) -> T:
        """访问位置i的元素"""
        if i < 0 or i >= self._cur_length:
            raise OutOfRange()
        
        p = self._head.next
        for count in range(i):
            if p is None:
                raise OutOfRange()
            p = p.next
            if p is None:
                raise OutOfRange()
        if p is None or p.data is None:
            raise OutOfRange()
        return p.data
    
    def inverse(self) -> None:
        """逆置链表，对应C++的算法"""
        p = self._head.next
        self._head.next = None
        
        if p is not None:
            self._tail = p  # 第一个节点成为尾节点
        
        # 头插法重建链表
        while p is not None:
            tmp = p.next
            p.next = self._head.next
            self._head.next = p
            p = tmp

# 双链表节点 - 对应C++的DoubleNode
class _DoubleLinkNode(Generic[T]):
    """双链表节点"""
    def __init__(self, data: Optional[T] = None, 
                 prior: Optional['_DoubleLinkNode[T]'] = None,
                 next_node: Optional['_DoubleLinkNode[T]'] = None):
        self.data = data  # 对应C++的data
        self.prior = prior  # 对应C++的prior指针
        self.next = next_node  # 对应C++的next指针

# 双链表实现 - 对应C++的doubleLinkList
class DoubleLinkList(List[T]):
    """双链表实现，使用头尾哨兵节点"""
    
    def __init__(self):
        """构造函数，创建头尾哨兵节点"""
        self._head = _DoubleLinkNode[T]()  # 头哨兵
        self._tail = _DoubleLinkNode[T]()  # 尾哨兵
        self._head.next = self._tail
        self._tail.prior = self._head
        self._cur_length = 0
    
    def _get_position(self, i: int) -> Optional[_DoubleLinkNode[T]]:
        """获取位置i的节点，对应C++的getPosition()"""
        if i < -1 or i > self._cur_length:
            return None
        
        p = self._head
        count = 0
        while count <= i and p is not None:
            p = p.next
            count += 1
        return p
    
    def clear(self) -> None:
        """清空双链表"""
        p = self._head.next
        self._head.next = self._tail
        self._tail.prior = self._head
        
        while p != self._tail and p is not None:
            tmp = p.next
            del p
            p = tmp
        
        self._cur_length = 0
    
    def empty(self) -> bool:
        """判断是否为空"""
        return self._head.next == self._tail
    
    def size(self) -> int:
        """返回链表长度"""
        return self._cur_length
    
    def traverse(self) -> None:
        """遍历输出双链表"""
        p = self._head.next
        while p != self._tail and p is not None:
            print(p.data, end=" ")
            p = p.next
        print()
    
    def search(self, value: T) -> int:
        """查找元素value"""
        p = self._head.next
        count = 0
        while p != self._tail and p is not None and p.data != value:
            p = p.next
            count += 1
        return count if p != self._tail else -1
    
    def insert(self, i: int, value: T) -> None:
        """在位置i插入元素value"""
        if i < 0 or i > self._cur_length:
            raise OutOfRange()
        
        p = self._get_position(i)
        if p is None or p.prior is None:
            raise OutOfRange()
        
        # 创建新节点并插入
        tmp = _DoubleLinkNode(value, p.prior, p)
        p.prior.next = tmp
        p.prior = tmp
        self._cur_length += 1
    
    def remove(self, i: int) -> None:
        """删除位置i的元素"""
        if i < 0 or i > self._cur_length - 1:
            raise OutOfRange()
        
        p = self._get_position(i)
        if p is None or p == self._tail:
            raise OutOfRange()
        
        # 删除节点
        if p.prior is not None and p.next is not None:
            p.prior.next = p.next
            p.next.prior = p.prior
        del p
        self._cur_length -= 1
    
    def visit(self, i: int) -> T:
        """访问位置i的元素"""
        if i < 0 or i >= self._cur_length:
            raise OutOfRange()
        
        p = self._head.next
        for count in range(i):
            if p is None:
                raise OutOfRange()
            p = p.next
        
        if p is None or p.data is None:
            raise OutOfRange()
        return p.data
    
    def inverse(self) -> None:
        """逆置双链表，对应C++算法"""
        p = self._head.next
        self._head.next = self._tail
        self._tail.prior = self._head
        
        # 头插法重建链表
        while p != self._tail and p is not None:
            tmp = p.next
            p.next = self._head.next
            p.prior = self._head
            self._head.next.prior = p
            self._head.next = p
            p = tmp

# 测试示例 - 与C++版本完全一致
if __name__ == "__main__":
    try:
        # 顺序表测试
        sl = SeqList[int](5)
        for i in range(5):
            sl.insert(i, i*2)
        print("顺序表初始: ", end="")
        sl.traverse()
        sl.inverse()
        print("顺序表逆置: ", end="")
        sl.traverse()

        # 单链表测试
        ll = LinkList[int]()
        for i in range(5):
            ll.insert(i, i+1)
        print("单链表初始: ", end="")
        ll.traverse()
        ll.remove(2)
        print("删除第三个元素: ", end="")
        ll.traverse()

        # 双链表测试
        dl = DoubleLinkList[int]()
        for i in range(3):
            dl.insert(i, i*3)
        print("双链表初始: ", end="")
        dl.traverse()
        dl.inverse()
        print("双链表逆置: ", end="")
        dl.traverse()
        
    except (OutOfRange, BadSize) as e:
        print(f"异常: {e}")